### ISSUE TITLE
Running test from the command line using maven mvn compile test 

### ISSUE THREAD

**Issue by Sanderson-Imbeah** [TNG/junit-dataprovider#51]
I have notice that i seem to be getting an error:

INFO] Changes detected - recompiling the module!
WARNING] File encoding has not been set, using platform encoding Cp1252, i.e. b
ild is platform dependent!
INFO] Compiling 3 source files to C:\Users\921495\IdeaProjects\HMW POC\target\c
asses
INFO] -------------------------------------------------------------
ERROR] COMPILATION ERROR :
INFO] -------------------------------------------------------------
ERROR] /C:/Users/921495/IdeaProjects/HMW POC/src/main/java/com/dataprovider/Dat
Providers.java:[3,43] package com.tngtech.java.junit.dataprovider does not exis

ERROR] /C:/Users/921495/IdeaProjects/HMW POC/src/main/java/com/dataprovider/Dat
Providers.java:[11,6] cannot find symbol
 symbol:   class DataProvider
 location: class com.dataprovider.DataProviders
INFO] 2 errors
INFO] -------------------------------------------------------------
INFO] ------------------------------------------------------------------------
INFO] BUILD FAILURE
INFO] ------------------------------------------------------------------------
INFO] Total time: 2.370 s
INFO] Finished at: 2015-08-12T11:38:59+01:00
INFO] Final Memory: 18M/221M
INFO] ------------------------------------------------------------------------
ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.
:compile (default-compile) on project my-project-halo: Compilation failure: Com
ilation failure:
ERROR] /C:/Users/921495/IdeaProjects/HMW POC/src/main/java/com/dataprovider/Dat
Providers.java:[3,43] package com.tngtech.java.junit.dataprovider does not exis

ERROR] /C:/Users/921495/IdeaProjects/HMW POC/src/main/java/com/dataprovider/Dat
Providers.java:[11,6] cannot find symbol
ERROR] symbol:   class DataProvider
ERROR] location: class com.dataprovider.DataProviders
ERROR] -> [Help 1]
ERROR]
ERROR] To see the full stack trace of the errors, re-run Maven with the -e swit
h.
ERROR] Re-run Maven using the -X switch to enable full debug logging.
ERROR]
ERROR] For more information about the errors and possible solutions, please rea
 the following articles:
ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureExc
ption

Does mvn support the data provider?

**Comment by aaschmid** [comment]
Hi @Sanderson-Imbeah,
I don't know exactly what you are doing but it seems that your compile job searches at the wrong package for any dataprovider classes. You search in 

```src/main/java/com/dataprovider/DataProviders.java``` for a package ```com.tngtech.java.junit.dataprovider``` which is usually located in ```src/main/java/com/tngtech/java/junit/dataprovider/DataProviders.java```.

Btw, to avoid the warning: the source code of junit-dataprovider is compelte written in UTF-8 but I just use us-ascii compatible characters such that it should work if you set your encoding to UTF-8, see http://maven.apache.org/general.html#encoding-warning.
Cheers,
Andreas

**Comment by Sanderson-Imbeah** [comment]
Hi @aaschmid 

I have place the pom code for dataprovider into my pom.xml and in my project structure i have created a dataprovider class which is located in C:\Users\921495\IdeaProjects\HMW POC\src\main\java\com\dataprovider\DataProviders.java the test is located in C:\Users\921495\IdeaProjects\HMW POC\src\test\java\com\selenium\test\ManipluationTest.java when i run the test from cmd using mvn clean test the error occurs ??

Many Thanks

**Comment by aaschmid** [comment]
Ah ok, can you provide me the pom.xml?

**Comment by Sanderson-Imbeah** [comment]
Hi @aaschmid 

Please see below

<?xml version="1.0" encoding="UTF-8"?>

<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>


    <groupId>org.halo.hmw</groupId>
    <artifactId>my-project-halo</artifactId>
    <version>1.0-SNAPSHOT</version>


    <properties>
        <junit.version>4.12</junit.version>
        <surefire.version>2.18</surefire.version>
    </properties>


    <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-surefire-plugin</artifactId>
                <version>2.18</version>
                <configuration>
                    <parallel>methods</parallel>
                    <threadCount>3</threadCount>
                </configuration>
            </plugin>
        </plugins>

    </build>


    <dependencies>

    <dependency>
        <groupId>org.seleniumhq.selenium</groupId>
        <artifactId>selenium-java</artifactId>
        <version>2.46.0</version>
    </dependency>

        <dependency>
        <groupId>junit</groupId>
        <artifactId>junit</artifactId>
        <version>4.12</version>
    </dependency>


        <dependency>
            <groupId>org.seleniumhq.selenium</groupId>
            <artifactId>selenium-chrome-driver</artifactId>
            <version>2.45.0</version>
        </dependency>

        <dependency>
            <groupId>org.seleniumhq.selenium</groupId>
            <artifactId>selenium-ie-driver</artifactId>
            <version>2.46.0</version>
        </dependency>
        
        <dependency>
            <groupId>com.tngtech.java</groupId>
            <artifactId>junit-dataprovider</artifactId>
            <version>1.10.0</version>
            <scope>test</scope>
        </dependency>


    </dependencies>




</project>

**Comment by Sanderson-Imbeah** [comment]
@aaschmid If i run the test from with the IDE no problem from maven cmd then the error occurs?

**Comment by Sanderson-Imbeah** [comment]
@aaschmid it seems to be a MojoFailureException which on the Maven site says the following:
Unlike many other errors, this exception is not generated by the Maven core itself but by a plugin. As a rule of thumb, plugins use this error to signal a failure of the build because there is something wrong with the dependencies or sources of a project, e.g. a compilation or a test failure.
The concrete meaning of the exception depends on the plugin so please have a look at its documentation. The documentation for many common Maven plugins can be reached via our plugin index.

**Comment by aaschmid** [comment]
Hi @Sanderson-Imbeah, I will hopefully have time to have a closer look until the end of the week. Hopefully the issue is not too urgent, otherwise give me a hint!

**Comment by Sanderson-Imbeah** [comment]
Hi @aaschmid Thanks it is  urgent as when we run our tests in parallel it does not work using the cmd line. Even when we IDE the test wont run in parallel as junit needs to be run from cmd in order for the pom that controls the parallel dependency to work. Thank you in advance for all your support so far hopefully you can get to the bottom of this

**Comment by aaschmid** [comment]
Hi @Sanderson-Imbeah, 
just using the provided ```pom.xml``` above and putting same junit-dataprovider test in there works fine. I assume because of not using Selenium. As I have never used selenium before, is is possible, that you write a simple example test combining selenium and junit-dataprovider to simply call some public webpage and check for different labels, links or buttons on it?
That would help me a lot reproducing and verifying your problems (and would safe me a lot time to get into Selenium).
Thanks, Andreas

**Comment by Sanderson-Imbeah** [comment]
Hi @aaschmid 


Thanks for getting back to me please see the below the following code. Bare in mind I am calling the dataprovider class from separate  class:

package com.dataprovider;

import com.tngtech.java.junit.dataprovider.DataProvider;

/**
 * 
 */
public class Data {



    @DataProvider
    public static Object [][] validData(  ) {

        return new Object [] [] {

                {"My name is Sanderson"}

        };
    }
}


The test below will go to a webpage type in "My name is Sanderson" and assert that the text has been passed in correctly

@RunWith(DataProviderRunner.class)

public class ManipluationTest {


    public   WebDriver driver;


    @Before
    public void setUp() {


        driver = new  FirefoxDriver ();
        
        driver.get("http://compendiumdev.co.uk/selenium/basic_html_form.html");
    }



    @After
    public  void tearDown() {
        driver.close();
    }
  @Test
    @UseDataProvider(value = "validData", location = Data.class)
    public void clearThenTypeCommentsSubmitFormAndCheckOutputTest(String text) {

        WebElement commentsbox = driver.findElement(By.cssSelector("[name = 'comments' ]"));

        commentsbox.clear();

        commentsbox.sendKeys(text);

        WebElement submit = driver.findElement(By.cssSelector("[type ='submit']"));

        submit.click();

        WebDriverWait wait = new WebDriverWait(driver, 700);

    wait.until(ExpectedConditions.urlToBe("http://compendiumdev.co.uk/selenium/form_processor.php"));

        WebElement commentsfield = driver.findElement(By.cssSelector("#_valuecomments"));

        String commentsentry = commentsfield.getText();


        assertThat(commentsentry, is("My name is Sanderson"));


    }


Hope this helps

**Comment by Sanderson-Imbeah** [comment]
@aaschmid I am also using intelliJ as the IDE.

**Comment by Sanderson-Imbeah** [comment]
@aaschmid Hi hope all is well have you managed to reproduce the error ?

Kind Regards

**Comment by aaschmid** [comment]
Hi @Sanderson-Imbeah, unfortunately your little reproducer is working on my machine (Archlinux) using the reproducer in https://github.com/aaschmid/junit-dataprovider-and-selenium. 

I am not sure, though, if they really run in parallel. Is running selenium tests in parallel not a bit dangerous as multiple browser instances has to be used concurrently? But that's another issue.
https://github.com/aaschmid/junit-dataprovider-and-selenium

**Comment by Sanderson-Imbeah** [comment]
@aaschmid Hi thanks for getting back to that in your pom file you the following 

 <dependency>
            <groupId>xml-apis</groupId>
            <artifactId>xml-apis</artifactId>
            <version>1.4.01</version>
        </dependency>


I didnt have this in my pom file when conducting the test (not sure if this will make a change)

Ok i will have another look at it and let you know of the outcome.

**Comment by aaschmid** [comment]
I got some ```NoClassDefFoundException```s while not having ```xml-apis``` as dependency.

**Comment by Sanderson-Imbeah** [comment]
@aaschmid I have notice that from the command line if enter mvn clean test then it fails while if i just use mvn compile test its fine?

**Comment by aaschmid** [comment]
Therefore it seems to be a problem with the ```clean``` task as ```compile``` will also be executed if you just use ```mvn test```. That's strange, though! Can you answer the following questions?

* What is the exception using ```mvn clean test```?
* Is it working if you do two separate commands for ```mvn clean``` and ```mvn test```?
* What is if you execute ```mvn clean package``` or ```mvn clean install```?

**Comment by Sanderson-Imbeah** [comment]
@aaschmid on the mvn clean test i get the following errors:

[ERROR] /C:/Users/921495/IdeaProjects/HMW POC/src/main/java/com/dataprovider/Dat
a.java:[3,43] package com.tngtech.java.junit.dataprovider does not exist
[ERROR] /C:/Users/921495/IdeaProjects/HMW POC/src/main/java/com/dataprovider/Dat
a.java:[12,6] cannot find symbol
  symbol:   class DataProvider
  location: class com.dataprovider.Data
[INFO] 2 errors
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 2.996 s
[INFO] Finished at: 2015-08-27T12:27:03+01:00
[INFO] Final Memory: 19M/221M

**Comment by Sanderson-Imbeah** [comment]
on mvn clean the build is successfully however i get the same error once i use mvn test

**Comment by Sanderson-Imbeah** [comment]
mvn clean package also gives compilation errors

**Comment by Sanderson-Imbeah** [comment]
weird thing is if i run the test direct from the ide then i do mvn test from the command no more error

**Comment by aaschmid** [comment]
The problem must be with the building of your project as ```mvn clean``` deletes your ```target/``` folder and the IDE as well as the above mentioned command line ```mvn``` commands besides ```clear``` will compile your ```*.java``` files and refill your ```target/``` dir.
Even if I think that this problem is caused by the junit-dataprovider, can you update the above provided reproducer such that I can reproduce it?

**Comment by Sanderson-Imbeah** [comment]
@aaschmid  i have placed the pom that i am using as of now i really do think the problem is with junit provider

**Comment by Sanderson-Imbeah** [comment]
INFO] Copying 3 resources
INFO]
INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ my-project-halo
---
INFO] Changes detected - recompiling the module!
WARNING] File encoding has not been set, using platform encoding Cp1252, i.e. b
ild is platform dependent!
INFO] Compiling 3 source files to C:\Users\921495\IdeaProjects\HMW POC\target\c
asses
INFO] -------------------------------------------------------------
ERROR] COMPILATION ERROR :
INFO] -------------------------------------------------------------
ERROR] /C:/Users/921495/IdeaProjects/HMW POC/src/main/java/com/dataprovider/Dat
.java:[3,43] package com.tngtech.java.junit.dataprovider does not exist
ERROR] /C:/Users/921495/IdeaProjects/HMW POC/src/main/java/com/dataprovider/Dat
.java:[11,6] cannot find symbol
 symbol:   class DataProvider
 location: class com.dataprovider.Data
INFO] 2 errors
INFO] -------------------------------------------------------------
INFO] ------------------------------------------------------------------------
INFO] BUILD FAILURE
INFO] ------------------------------------------------------------------------
INFO] Total time: 2.182 s
INFO] Finished at: 2015-08-28T12:43:15+01:00
INFO] Final Memory: 18M/150M
INFO] ------------------------------------------------------------------------
ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.
:compile (default-compile) on project my-project-halo: Compilation failure: Com
ilation failure:
ERROR] /C:/Users/921495/IdeaProjects/HMW POC/src/main/java/com/dataprovider/Dat
.java:[3,43] package com.tngtech.java.junit.dataprovider does not exist
ERROR] /C:/Users/921495/IdeaProjects/HMW POC/src/main/java/com/dataprovider/Dat
.java:[11,6] cannot find symbol
ERROR] symbol:   class DataProvider
ERROR] location: class com.dataprovider.Data
ERROR] -> [Help 1]
ERROR]
ERROR] To see the full stack trace of the errors, re-run Maven with the -e swit
h.
ERROR] Re-run Maven using the -X switch to enable full debug logging.
ERROR]
ERROR] For more information about the errors and possible solutions, please rea
 the following articles:
ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureExc
ption

:\Users\921495\IdeaProjects\HMW POC>

**Comment by Sanderson-Imbeah** [comment]
Why when i clean from the ide it works fine ?? also the which os are you using ide etc etc

**Comment by aaschmid** [comment]
Wait, wait, wait. With the above pasted ```pom.xml``` and your now pasted complete ```mvn``` log output, the problem is most likely the scoping of the junit-dataprovider dependency. You have a compile error which says that the junit-dataprovider packages cannot be found. And this is correct because your ```pom.xml``` scopes your junit-dataprovider dependency to ```<scope>test</scope>```. Does that fix your problem (as it is very unclear to me why it works on the second run if clean is omitted)? 

And have you ever tried to run the reproducer? I suppose not because the pom.xml has leading newlines and spaces which made the build fail before it has started at all. I have fixed the pom and again get the ```java.lang.NoClassDefFoundError: org/w3c/dom/ElementTraversal``` within the IDE. On the command line a ```java.lang.NullPointerException``` occurs which is most likely caused by the above ```java.lang.NoClassDefFoundError``` (maven just hides this anyhow, though).

**Comment by Sanderson-Imbeah** [comment]
Let me try and change the scope of my Pom.xml and see if that fixes the issue.

I did run the reproducer but got an error lol let me try this and get back to you.

**Comment by Sanderson-Imbeah** [comment]
It seems to be work fine now after I rebuilt the project.

**Comment by aaschmid** [comment]
Great!
